### PR TITLE
DfTx: `setGovHeight`

### PR DESCRIPTION
**Pull request by ivan-zynesis**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
New DfTx type in FC

#### Which issue(s) does this PR fixes?:

Fixes part of #532 

#### Additional comments?:
Not txn-builder is added as it is only for foundation (same for `setGov`), solely for decoding purpose.

### CODE REVIEW

**Review by surangap — COMMENTED** [review]
1. need to add transaction builder tests?
2. Code for `CGovernanceVar` composers is not that elegant🤔. But I guess we can not do much about it without changing the way how cpp side handles gov vars.

**Comment by ivan-zynesis** [comment]
> 1. need to add transaction builder tests?
> 2. Code for `CGovernanceVar` composers is not that elegant🤔. But I guess we can not do much about it without changing the way how cpp side handles gov vars.

1. No transaction-builder is added at all thus no test. Consider the `CSetGovernanceVar` wasn't there either. The jellyfish-txn is mainly for scan or other client to decode/read existing dftx history.
2. Yes, cannot really make changes, there is no GovVar length. Also quite impossible to change for sake of backward compatibility.

**Review by canonbrother — DISMISSED** [review]

**Comment by surangap** [comment]
> > 1. need to add transaction builder tests?
> > 2. Code for `CGovernanceVar` composers is not that elegant🤔. But I guess we can not do much about it without changing the way how cpp side handles gov vars.
> 
> 1. No transaction-builder is added at all thus no test. Consider the `CSetGovernanceVar` wasn't there either. The jellyfish-txn is mainly for scan or other client to decode/read existing dftx history.
> 2. Yes, cannot really make changes, there is no GovVar length. Also quite impossible to change for sake of backward compatibility.

so for number 1, your point is that since only the foundation sets gov vars, jellyfish does not need it?

**Comment by ivan-zynesis** [comment]
> > > 1. need to add transaction builder tests?
> > > 2. Code for `CGovernanceVar` composers is not that elegant🤔. But I guess we can not do much about it without changing the way how cpp side handles gov vars.
> > 
> > 
> > 
> > 1. No transaction-builder is added at all thus no test. Consider the `CSetGovernanceVar` wasn't there either. The jellyfish-txn is mainly for scan or other client to decode/read existing dftx history.
> > 2. Yes, cannot really make changes, there is no GovVar length. Also quite impossible to change for sake of backward compatibility.
> 
> so for number 1, your point is that since only the foundation sets gov vars, jellyfish does not need it for setting?

The implementation in this PR at the moment cover only `jellyfish-transaction` but not `jellyfish-transaction-builder` thus no test for txn-builder

We can/should add the builder if there is need to perform `setGov` using jellyfish library together with foundation auth. If that's the case, we will have to add `setGov` dftx builder as well. Is the need foreseen? What do you think? @surangap @fuxingloh

**Comment by fuxingloh** [comment]
I am in the middle, with builder test at least any basic test, we can validate the authenticity and safeguard ourself from regression. It's also the only way to know it's correct and can be submitted to chain now.

Many time it was useful in migration where upstream altered the structure without informing.

I guess it's also forward looking as we want to eventually achieve full node compatibility using N-API?

I guess we can meet in the middle now with extremely basic positive test to submit?

**Comment by ivan-zynesis** [comment]
Okay will add the txn builder tomorrow (should very quick one, literally a function calling super.createDfTx) and minimal test.

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_governance.ts [hunk 1]
```diff
@@ -59,4 +59,32 @@ export class TxnBuilderGovernance extends P2WPKHTxnBuilder {
...
+   */
+  async setGoverance (setGov: SetGovernance, changeScript: Script): Promise<TransactionSegWit> {
+    return await this.createDeFiTx(
+      OP_CODES.OP_DEFI_TX_SET_GOVERNANCE(setGov),
+      changeScript
+    )
+  }
+
+  /**
+   * Set governance variable.
```
Comment (surangap) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder_governance.ts:78]:
```suggestion
   * Set governance variable at the given height
```

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_governance.ts [hunk 2]
```diff
@@ -59,4 +59,32 @@ export class TxnBuilderGovernance extends P2WPKHTxnBuilder {
...
+   *
+   * @param {SetGovernance} setGov txn to create
```
Comment (surangap) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder_governance.ts:80]:
```suggestion
      * @param {SetGovernanceHeights} setGovHeight txn to create
```

**Review by surangap — COMMENTED** [review]
looks good. minor comments. Also please add the non foundation failure test.

**Review by surangap — APPROVED** [review]

**Review by jingyi2811 — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**